router/batchrouter/asyncdestinationmanager/salesforce-bulk/salesforce_bulk.go

```diff
@@ -0,0 +1,460 @@
+	var sfJobs []SalesforceJobInfo
+
+	allFailedJobIDs = append(allFailedJobIDs, failedJobIDs...)
+
+	for len(input) > 0 {
+		s.hashMapMutex.Lock()
+		csvFilePath, csvHeaders, insertedJobIDs, overflowedJobs, err := createCSVFile(
+			destinationID,
+			input,
+			s.dataHashToJobID,
+		)
+		s.hashMapMutex.Unlock()
+
+		if err != nil {
+			s.logger.Errorn("Error creating CSV: %v", obskit.Error(err))
+			for _, job := range input {
+				if jobID, ok := job.Metadata["job_id"].(float64); ok {
+					allFailedJobIDs = append(allFailedJobIDs, int64(jobID))
+				}
+			}
+			break
+		}
+
+		if len(insertedJobIDs) == 0 {
+			if err := os.Remove(csvFilePath); err != nil {
+				s.logger.Debugn("Failed to remove empty CSV file %s: %v", logger.NewStringField("csvFilePath", csvFilePath), obskit.Error(err))
+			}
+			s.logger.Errorn("No jobs fit in CSV, marking as failed")
+			for _, job := range input {
+				if jobID, ok := job.Metadata["job_id"].(float64); ok {
+					allFailedJobIDs = append(allFailedJobIDs, int64(jobID))
+				}
+			}
+			break
+		}
+
+		s.logger.Infon("Created CSV with %d jobs (batch %d of %d total)",
+			logger.NewIntField("jobs", int64(len(insertedJobIDs))),
+			logger.NewIntField("batch", int64(len(allImportingJobIDs)/100+1)),
+			logger.NewIntField("total", int64(len(input))),
+		)
+		sfJobID, apiError := s.apiService.CreateJob(
+			objectInfo.ObjectType,
+			"upsert",
+			objectInfo.ExternalIDField,
+		)
+		if apiError != nil {
+			s.logger.Errorn("Error creating Salesforce job for operation upsert: %v", logger.NewStringField("apiErrorMessage", apiError.Message))
+			allFailedJobIDs = append(allFailedJobIDs, insertedJobIDs...)
+			for _, job := range overflowedJobs {
+				allFailedJobIDs = append(allFailedJobIDs, int64(job.Metadata["job_id"].(float64)))
+			}
+			break
+		}
+
+		apiError = s.apiService.UploadData(sfJobID, csvFilePath)
+		if apiError != nil {
+			s.logger.Errorn("Error uploading data for operation upsert: %v", logger.NewStringField("apiErrorMessage", apiError.Message))
+			_ = s.apiService.DeleteJob(sfJobID)
+			allFailedJobIDs = append(allFailedJobIDs, insertedJobIDs...)
+			for _, job := range overflowedJobs {
+				allFailedJobIDs = append(allFailedJobIDs, int64(job.Metadata["job_id"].(float64)))
+			}
+			break
+		}
+
+		apiError = s.apiService.CloseJob(sfJobID)
+		if apiError != nil {
+			s.logger.Errorn("Error closing job for operation upsert: %v", logger.NewStringField("apiErrorMessage", apiError.Message))
+			allFailedJobIDs = append(allFailedJobIDs, insertedJobIDs...)
+			for _, job := range overflowedJobs {
+				allFailedJobIDs = append(allFailedJobIDs, int64(job.Metadata["job_id"].(float64)))
+			}
+			break
+		}
+
+		s.logger.Infon("Successfully created and closed Salesforce Bulk job %s for operation upsert", logger.NewStringField("jobID", sfJobID))
+
+		allImportingJobIDs = append(allImportingJobIDs, insertedJobIDs...)
+		sfJobs = append(sfJobs, SalesforceJobInfo{
+			ID:      sfJobID,
+			Headers: csvHeaders,
```

Comment:
Why are we storing headers as part of info? Also, why are we storing jobs twice as part of importInfo?

```GO
	importData := map[string]interface{}{"jobs": sfJobs}
	importID, _ := jsonrs.Marshal(importData)

	paramsWithImportID := map[string]interface{}{
		"importId": string(importID),
		"jobs":     sfJobs,
	}
	jobsJSON, _ := jsonrs.Marshal(paramsWithImportID)
```